blog/2024-03-15-reflex-architecture.md

```diff
@@ -0,0 +1,166 @@
+
+When building Reflex, we wanted the final product to look and feel like a traditional web app to the end user, while still being easy to build and maintain for the developer. We chose to leverage the existing web ecosy
+
+The diagram below shows a detailed overview of how a Reflex app works.
+We'll go through each part in more detail in the following sections.
+
+![Reflex Architecture](/architecture.png)
+
+```python eval
+rx.box(height="1em")
+```
+
+### Frontend (Components)
+
+Reflex frontends are built using components that can be composed together to create complex UIs.
+
+```python
+def index():
+    return rx.hstack(
+        rx.avatar(src=f"https://github.com/{GithubState.username}.png"),
+        rx.input(
+            placeholder="Enter your username",
+            on_blur=GithubState.set_username,
+        ),
+    )
+```
+
+In our example app, we have components such as `rx.hstack`, `rx.avatar`, and `rx.input`. Under the hood, these components compile down to React components.
+
+For example, the above code compiles down to the following React code:
+
+```jsx
+<HStack>
+    <Avatar src="https://github.com/{GithubState.username}.png">
+    <Input
+        placeholder="Enter your username"
+        onBlur={GithubState.set_username}
+    >
+</HStack>
+```
+
+We chose React because it is a popular library with a huge ecosystem. Our goal isn't to recreate the web ecosystem, but to make it accessible to Python developers.
+
+This also let's our users bring their own components if we don't have a component they need.
+
+### Backend (State)
+
+In Reflex only the frontend compiles to Javascript, while all the state and logic stays in Python and is run on the server.
+
+```python
+import reflex as rx
+
+class GithubState(rx.State):
+    username: str = "reflex-dev"
+
+    def set_username(self, username: str):
+        self.username = username
+```
+
+Normally when writing web apps, you have to write a lot of boilerplate code to connect the frontend and backend. With Reflex, you don't have to worry about that. We handle the communication between the frontend and backend for you.
+
+We follow a reactive approach to state. All the variables that change over time are defined in a `State` class. When a variable changes, the frontend is automatically updated to reflect the new state.
+
+
+## Event Processing
+
+Now we get into the interesting part - how we handle events and state updates. Let's walk through it with our Github profile image example.
+
+### Event Trigger
+
+Events are initiated when the user interacts with the UI. In our example, when the user types in the input field and then clicks away, the `on_blur` event is triggered.
+
+### Event Queue
+
```

Comment:
here we mention an `event handler` for the first time but we did not define what it is exactly